File: apps/ios/app/Services/SmartSuggestionsService.swift
import Foundation

// MARK: - SmartSuggestionsService

/// Analyzes past sessions to suggest object types when creating a new session.
///
/// Uses frequency analysis across all sessions to recommend the most commonly
/// used object type names, colors, and icons.
///
/// This feature gives OpenCount an advantage over ZapCount and CountThings
/// which require users to manually define all object types every time.
final class SmartSuggestionsService {

    // MARK: - Suggestion model

    struct ObjectTypeSuggestion: Identifiable, Hashable {
        let id = UUID()
        let name: String
        let colorHex: String
        let iconName: String
        let usageCount: Int
        let lastUsed: Date
    }

    // MARK: - Generate suggestions

    /// Returns the top N most frequently used object type names across all sessions,
    /// sorted by usage frequency descending.
    ///
    /// - Parameters:
    ///   - sessions: All available sessions to analyze.
    ///   - limit: Maximum number of suggestions to return.
    ///   - excludingNames: Names to exclude (e.g. already added to current session).
    func suggestions(
        from sessions: [CountSession],
        limit: Int = 10,
        excludingNames: Set<String> = []
    ) -> [ObjectTypeSuggestion] {
        // Aggregate usage across all sessions
        var usageMap: [String: (count: Int, colorHex: String, iconName: String, lastUsed: Date)] = [:]

        for session in sessions {
            for objectType in session.objectTypes {
                let name = objectType.name
                guard !excludingNames.contains(name) else { continue }

                let existing = usageMap[name]
                let newCount = (existing?.count ?? 0) + 1
                let newLastUsed = max(existing?.lastUsed ?? .distantPast, session.modifiedAt)

                usageMap[name] = (
                    count: newCount,
                    colorHex: objectType.colorHex,
                    iconName: objectType.iconName,
                    lastUsed: newLastUsed
                )
            }
        }

        // Sort by usage count descending, then by last used date
        return usageMap
            .map { name, data in
                ObjectTypeSuggestion(
                    name: name,
                    colorHex: data.colorHex,
                    iconName: data.iconName,
                    usageCount: data.count,
                    lastUsed: data.lastUsed
                )
            }
            .sorted { lhs, rhs in
                if lhs.usageCount != rhs.usageCount {
                    return lhs.usageCount > rhs.usageCount
                }
                return lhs.lastUsed > rhs.lastUsed
            }
            .prefix(limit)
            .map { $0 }
    }

    // MARK: - Icon suggestions

    /// Returns SF Symbol names that are commonly used for counting apps,
    /// grouped by category.
    static let iconCategories: [(category: String, icons: [String])] = [
        ("Animals", [
            "bird.fill", "ant.fill", "tortoise.fill", "hare.fill",
            "fish.fill", "pawprint.fill", "ladybug.fill", "lizard.fill"
        ]),
        ("Nature", [
            "leaf.fill", "tree.fill", "flame.fill", "drop.fill",
            "cloud.fill", "sun.max.fill", "moon.fill", "snowflake"
        ]),
        ("Objects", [
            "car.fill", "bicycle", "airplane", "shippingbox.fill",
            "bag.fill", "cart.fill", "cube.fill", "cylinder.fill"
        ]),
        ("People", [
            "person.fill", "person.2.fill", "person.3.fill",
            "figure.walk", "figure.run", "figure.stand"
        ]),
        ("Shapes", [
            "circle.fill", "square.fill", "triangle.fill", "diamond.fill",
            "star.fill", "heart.fill", "bolt.fill", "flag.fill"
        ]),
        ("Science", [
            "atom", "cross.fill", "waveform", "chart.bar.fill",
            "microscope", "stethoscope", "pills.fill", "syringe.fill"
        ]),
    ]

    // MARK: - Color palette

    /// A curated palette of colors suitable for counting markers.
    static let colorPalette: [(name: String, hex: String)] = [
        ("Red",       "#E74C3C"),
        ("Orange",    "#E67E22"),
        ("Yellow",    "#F1C40F"),
        ("Green",     "#2ECC71"),
        ("Teal",      "#1ABC9C"),
        ("Blue",      "#3498DB"),
        ("Indigo",    "#5B6EAE"),
        ("Purple",    "#9B59B6"),
        ("Pink",      "#E91E8C"),
        ("Brown",     "#795548"),
        ("Gray",      "#95A5A6"),
        ("Dark",      "#34495E"),
    ]
}

File: packages/shared/src/commonMain/kotlin/com/opencount/shared/service/SmartSuggestionsService.kt
package com.opencount.shared.service

import com.opencount.shared.model.CountSession
import com.opencount.shared.model.ObjectType

/**
 * A suggested object type derived from historical usage patterns.
 * @property frequency how many sessions this type appeared in
 */
data class ObjectTypeSuggestion(
    val name: String,
    val colorHex: String,
    val iconName: String,
    val frequency: Int = 0,
)

/** Default set of SF Symbol icon names available for object types. */
val iconCategories: List<String> = listOf(
    "circle.fill", "square.fill", "triangle.fill", "star.fill",
    "heart.fill", "flag.fill", "tag.fill", "bookmark.fill",
    "bell.fill", "fire.fill", "bolt.fill", "leaf.fill",
    "flame.fill", "drop.fill", "moon.fill", "sun.max.fill",
    "car.fill", "bus.fill", "bicycle", "airplane",
    "house.fill", "building.fill", "cart.fill", "bag.fill",
)

/** Default palette of hex color strings available for object types. */
val colorPalette: List<String> = listOf(
    "#FF5733", "#33FF57", "#3357FF", "#FF33F5",
    "#33FFF5", "#F5FF33", "#FF8333", "#8333FF",
    "#33FF83", "#FF3383", "#3383FF", "#83FF33",
)

/**
 * Service that suggests object types based on historical usage frequency across sessions.
 * Also provides a set of default types for new sessions.
 */
class SmartSuggestionsService {
    /**
     * Returns the most frequently used object types across [sessions], optionally excluding certain names.
     * @param limit maximum number of suggestions to return
     * @param excludingNames type names to exclude from results
     * @return list of suggestions sorted by frequency descending
     */
    fun suggestions(
        sessions: List<CountSession>,
        limit: Int = 5,
        excludingNames: Set<String> = emptySet(),
    ): List<ObjectTypeSuggestion> {
        val frequencyMap = mutableMapOf<String, Int>()
        val typeMap = mutableMapOf<String, ObjectType>()

        for (session in sessions) {
            for (type in session.objectTypes) {
                if (type.name in excludingNames) continue
                frequencyMap[type.name] = (frequencyMap[type.name] ?: 0) + 1
                if (!typeMap.containsKey(type.name)) typeMap[type.name] = type
            }
        }

        return frequencyMap.entries
            .sortedByDescending { it.value }
            .take(limit)
            .map { (name, freq) ->
                val ot = typeMap[name]
                ObjectTypeSuggestion(
                    name = name,
                    colorHex = ot?.colorHex ?: colorPalette[0],
                    iconName = ot?.iconName ?: iconCategories[0],
                    frequency = freq,
                )
            }
    }

    /** Returns a list of four default [ObjectType]s (Cars, People, Trees, Animals) for new sessions. */
    fun generateDefaultTypes(): List<ObjectType> {
        return listOf(
            ObjectType.create("Cars", colorHex = "#FF5733"),
            ObjectType.create("People", colorHex = "#33FF57"),
            ObjectType.create("Trees", colorHex = "#3357FF"),
            ObjectType.create("Animals", colorHex = "#FF33F5"),
        )
    }
}
